File: Scalemon.SignalBus/SignalBus.vb
Imports Microsoft.Extensions.Configuration
Imports Microsoft.Extensions.Logging
Imports System.Drawing
Imports System.IO
Imports System.IO.Ports
Imports System.Timers
Imports Scalemon.Common
Public Class SignalBus
    Implements ISignalBus, IDisposable

    Private ReadOnly _buttonpressedHandlers As New List(Of Func(Of Task))()

    Private _serialPort As SerialPort
    Private _portName As String = "COM1"
    Private _baudRate As Integer = 9600
    Private _reconnectIntervalMs As Integer = 5000
    Private _timer As Timer
    Private _isConnected As Boolean = False
    Private disposedValue As Boolean
    Private ReadOnly _config As IConfiguration
    Private ReadOnly _logger As ILogger(Of SignalBus)
    Public Sub New(config As IConfiguration, logger As ILogger(Of SignalBus))
        _config = config
        _logger = logger
    End Sub

    Public Property PortName As String Implements ISignalBus.PortName
        Get
            Return _portName
        End Get
        Set(value As String)
            _portName = value
        End Set
    End Property

    Public Property BaudRate As Integer Implements ISignalBus.BaudRate
        Get
            Return _baudRate
        End Get
        Set(value As Integer)
            _baudRate = value
        End Set
    End Property

    Public Property ReconnectIntervalMs As Integer Implements ISignalBus.ReconnectIntervalMs
        Get
            Return _reconnectIntervalMs
        End Get
        Set(value As Integer)
            _reconnectIntervalMs = value
        End Set
    End Property


    Public Event ConnectionEstablished() Implements ISignalBus.ConnectionEstablished
    Public Event ConnectionLost() Implements ISignalBus.ConnectionLost





    Public Sub Start() Implements ISignalBus.Start
        _portName = _config("PLCSettings:PortName")
        _baudRate = _config("PLCSettings:BaudRate")
        _reconnectIntervalMs = _config("PLCSettings:ReconnectIntervalMs")
        _serialPort = New SerialPort()
        With _serialPort
            .PortName = _portName
            .BaudRate = _baudRate
            .Parity = Parity.None
            .DataBits = 8
            .StopBits = StopBits.One
            .ReadTimeout = 500
        End With

        AddHandler _serialPort.DataReceived, AddressOf OnDataReceived
        TryOpenPort()
        _timer = New Timer(ReconnectIntervalMs)
        AddHandler _timer.Elapsed, Sub(sender, e) If Not _isConnected Then TryOpenPort()
        _timer.Start()
    End Sub

    Public Sub [Stop]() Implements ISignalBus.Stop
        RemoveHandler _serialPort.DataReceived, AddressOf OnDataReceived
        If _serialPort IsNot Nothing AndAlso _serialPort.IsOpen Then
            _serialPort.Close()
        End If
        _timer?.Stop()
    End Sub

    Private Sub TryOpenPort()
        Try
            If Not _serialPort.IsOpen Then _serialPort.Open()
            If Not _isConnected Then
                _isConnected = True
                RaiseEvent ConnectionEstablished()
            End If
        Catch
            If _isConnected Then
                _isConnected = False
                RaiseEvent ConnectionLost()
            End If
        End Try
    End Sub

    Private Async Sub OnDataReceived(sender As Object, e As SerialDataReceivedEventArgs)
        Try
            Dim data As Integer = _serialPort.ReadByte()
            If data = &H20 Then ' сигнал кнопки
                Await RaiseAllAsync(_buttonpressedHandlers)
            End If
        Catch ex As Exception
            ' возможная потеря соединения
            _logger.LogError(ex, "Ошибка при опросе весов")
        End Try
    End Sub

    Protected Overridable Sub Dispose(disposing As Boolean)
        If Not disposedValue Then
            If disposing Then
                ' 1) Останавливаем логику Stop()
                Me.Stop()
                ' 2) Освобождаем SerialPort
                If _serialPort IsNot Nothing Then
                    _serialPort.Dispose()
                    _serialPort = Nothing
                End If
                ' 3) Освобождаем Timer
                If _timer IsNot Nothing Then
                    _timer.Dispose()
                    _timer = Nothing
                End If
            End If
            ' === Здесь можно освободить неуправляемые ресурсы, если бы они были ===
            disposedValue = True
        End If
    End Sub

    Public Sub Dispose() Implements IDisposable.Dispose
        ' Не изменяйте этот код. Разместите код очистки в методе "Dispose(disposing As Boolean)".
        Dispose(disposing:=True)
        GC.SuppressFinalize(Me)
    End Sub

    Protected Overrides Sub Finalize()
        ' в финализаторе освобождаем только неуправляемые
        Dispose(disposing:=False)
        MyBase.Finalize()
    End Sub

    Private Async Function RaiseAllAsync(Of T)(handlers As IEnumerable(Of Func(Of T, Task)), arg As T) As Task
        For Each h In handlers.ToArray()
            Await h(arg)
        Next
    End Function

    Private Async Function RaiseAllAsync(handlers As IEnumerable(Of Func(Of Task))) As Task
        For Each h In handlers.ToArray()
            Await h()
        Next
    End Function

    Public Sub SubscribeButtonPressed(handler As Func(Of Task)) Implements ISignalBus.SubscribeButtonPressed
        _buttonpressedHandlers.Add(handler)
    End Sub

    Public Sub UnsubscribeButtonPressed(handler As Func(Of Task)) Implements ISignalBus.UnsubscribeButtonPressed
        _buttonpressedHandlers.Remove(handler)
    End Sub

    Public Async Function SendAsync(cmd As ArduinoSignalCode) As Task Implements ISignalBus.SendAsync
        If _serialPort.IsOpen Then
            Await Task.Run(Sub()
                               _serialPort.Write(New Byte() {CByte(cmd)}, 0, 1)
                           End Sub)
        End If
    End Function
End Class

File: Scalemon.SignalBus/SignalBus.cs
using System;
using System.Collections.Generic;
using System.IO.Ports;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.Extensions.Logging;
using Scalemon.Common;
using Timer = System.Timers.Timer;
#nullable enable
namespace Scalemon.SignalBus
{
    public class SignalBus : ISignalBus, IDisposable
    {
        private readonly List<Func<Task>> _buttonpressedHandlers = new List<Func<Task>>();
        private SerialPort? _serialPort;
        private readonly string _portName;
        private readonly int _baudRate;
        private readonly int _reconnectIntervalMs;
        private Timer? _timer;
        private bool _isConnected = false;
        private bool _openPortErrorLogged = false;
        private bool _openPortAttemptLogged = false;
        private readonly object _connectionGate = new();
        private readonly SemaphoreSlim _sendGate = new(1, 1);
        private bool disposedValue;
        private readonly ILogger<SignalBus> _logger;

        public SignalBus(ILogger<SignalBus> logger, string portName, int baudRate, int reconnectIntervalMs)
        {
            _logger = logger;
            _portName = portName;
            _baudRate = baudRate;
            _reconnectIntervalMs = reconnectIntervalMs;
            _logger.LogInformation(
                "SignalBus инициализирован: Порт={port}, Скорость={baud}, Интервал переподключения={interval}мс",
                _portName, _baudRate, _reconnectIntervalMs);
        }

        public event Action? ConnectionEstablished;
        
        public event Action? ConnectionLost;

        public event Action<SlaughterMode>? SlaughterModeChanged;

        public void Start()
        {
            _logger.LogInformation("Запуск SignalBus на порту {port}...", _portName);
            _serialPort = new SerialPort
            {
                PortName = _portName,
                BaudRate = _baudRate,
                Parity = Parity.None,
                DataBits = 8,
                StopBits = StopBits.One,
                ReadTimeout = 500
            };

            _serialPort.DataReceived += OnDataReceived;
            TryOpenPort();
            _timer = new Timer(_reconnectIntervalMs);
            _timer.Elapsed += (sender, e) => { if (!Volatile.Read(ref _isConnected)) TryOpenPort(); };
            _timer.Start();
        }

        public void Stop()
        {
            _logger.LogInformation("Остановка SignalBus.");
            if (_serialPort != null)
            {
                _serialPort.DataReceived -= OnDataReceived;
                if (_serialPort.IsOpen)
                {
                    _serialPort.Close();
                }
            }
            _timer?.Stop();
        }

        private void TryOpenPort()
        {
            try
            {
                if (_serialPort == null || _serialPort.IsOpen) return;

                if (!_openPortAttemptLogged)
                {
                    _logger.LogDebug("Попытка открыть порт {port}...", _portName);
                    _openPortAttemptLogged = true;
                }
                _serialPort.Open();

                var notifyConnected = false;
                lock (_connectionGate)
                {
                    if (!_isConnected)
                    {
                        _isConnected = true;
                        notifyConnected = true;
                    }
                }
                if (notifyConnected)
                {
                    _logger.LogInformation("Соединение с Arduino на порту {port} установлено.", _portName);

                    // Сбрасываем флаг ошибки, так как мы успешно подключились
                    _openPortErrorLogged = false; // <-- ДОБАВЛЕНО
                    _openPortAttemptLogged = false;

                    ConnectionEstablished?.Invoke();
                }
            }
            catch (Exception ex)
            {
                if (!MarkDisconnected(ex))
                {
                    // --- НАЧАЛО ИЗМЕНЕНИЙ ---
                    // Пишем в лог только если мы еще не сообщали об этой ошибке
                    if (!_openPortErrorLogged)
                    {
                        _logger.LogWarning("Не удалось открыть порт {port}: {message}. Следующая попытка через {interval} мс.", _portName, ex.Message, _reconnectIntervalMs);
                        // Ставим флаг, чтобы больше не писать об этом до успешного подключения
                        _openPortErrorLogged = true;
                    }
                    // --- КОНЕЦ ИЗМЕНЕНИЙ ---
                }
            }
        }

        private async void OnDataReceived(object sender, SerialDataReceivedEventArgs e)
        {
            try
            {
                if (_serialPort == null) return;
                while (_serialPort.BytesToRead > 0)
                {
                    int data = _serialPort.ReadByte();
                    if (data == 0x20) // сигнал кнопки
                    {
                        _logger.LogInformation("Получен сигнал нажатия кнопки (0x20) от Arduino.");
                        await RaiseAllAsync(_buttonpressedHandlers);
                    }
                    else if (data == 0x21)
                    {
                        SlaughterModeChanged?.Invoke(SlaughterMode.General);
                    }
                    else if (data == 0x22)
                    {
                        SlaughterModeChanged?.Invoke(SlaughterMode.Sanitary);
                    }
                    else
                    {
                        _logger.LogDebug("Получены неопознанные данные от Arduino: 0x{data:X2}", data);
                    }
                }
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Ошибка при чтении данных от Arduino");
                MarkDisconnected(ex);
            }
        }

        public async Task SendAsync(Enums.ArduinoSignalCode cmd)
        {
            await _sendGate.WaitAsync();
            try
            {
                if (_serialPort?.IsOpen == true)
                {
                    _logger.LogInformation("Отправка команды на Arduino: {command} (0x{commandCode:X2})", cmd, (byte)cmd);
                    await Task.Run(() => _serialPort.Write(new byte[] { (byte)cmd }, 0, 1));
                }
                else if (!MarkDisconnected())
                {
                    _logger.LogDebug("Команда {command} не отправлена: порт {port} ещё не открыт.", cmd, _portName);
                }
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Ошибка при отправке команды {command} на Arduino", cmd);
                MarkDisconnected(ex);
            }
            finally
            {
                _sendGate.Release();
            }
        }

        private bool MarkDisconnected(Exception? exception = null)
        {
            var notifyLost = false;
            lock (_connectionGate)
            {
                if (_isConnected)
                {
                    _isConnected = false;
                    notifyLost = true;
                }
            }

            try
            {
                if (_serialPort?.IsOpen == true)
                {
                    _serialPort.Close();
                }
            }
            catch (Exception closeException)
            {
                _logger.LogDebug(closeException, "Не удалось закрыть повреждённый порт {port}.", _portName);
            }

            if (notifyLost)
            {
                _logger.LogWarning(
                    exception,
                    "Потеряно соединение с Arduino на порту {port}.",
                    _portName);
                _openPortAttemptLogged = false;
                ConnectionLost?.Invoke();
            }

            return notifyLost;
        }

        #region Boilerplate Code (Subscribe/Dispose)
        public void SubscribeButtonPressed(Func<Task> handler) => _buttonpressedHandlers.Add(handler);
        public void UnsubscribeButtonPressed(Func<Task> handler) => _buttonpressedHandlers.Remove(handler);
        private async Task RaiseAllAsync(IEnumerable<Func<Task>> handlers)
        {
            foreach (var h in handlers.ToArray()) await h();
        }

        // ... Остальной код (Dispose, деструктор и т.д.) без изменений ...
        protected virtual void Dispose(bool disposing)
        {
            if (!disposedValue)
            {
                if (disposing)
                {
                    Stop();
                    if (_serialPort is not null)
                    {
                        _serialPort.Dispose();
                        _serialPort = null;
                    }
                    if (_timer is not null)
                    {
                        _timer.Dispose();
                        _timer = null;
                    }
                }
                disposedValue = true;
            }
        }

        public void Dispose()
        {
            Dispose(disposing: true);
            GC.SuppressFinalize(this);
        }

        ~SignalBus()
        {
            Dispose(disposing: false);
        }
        #endregion
    }
}
